src/date.c

```diff
@@ -1,18 +1,981 @@
+}
+
+static etimeint_t ecma_TimeFromYear(etimeint_t y) {
+  return msPerDay * ecma_DayFromYear(y);
+}
+
+static int ecma_YearFromTime_s(etime_t t)
+{
+  int first = floor((t / msPerDay) / 366) + 1970,
+              last = floor((t / msPerDay) / 365) + 1970,
+              middle = 0;
+  
+  if (last < first) {
+    int temp = first;
+    first = last;
+    last = temp;
+  }
+  
+  while (last > first) {
+    middle = (last + first) / 2;
+    if (ecma_TimeFromYear(middle) > t) {
+      last = middle - 1;
+    } else {
+      if (ecma_TimeFromYear(middle) <= t) {
+        if (ecma_TimeFromYear(middle + 1) > t) {
+          first = middle;
+          break;
+        }
+        first = middle + 1;
+      }
+    }
+  }
+  
+  return first;
+}
+
+static int ecma_IsLeapYear(int year) {
+  return ecma_DaysInYear(year) == 366;
+}
+
+static int ecma_DayWithinYear(etime_t t, int year) {
+  return (int)(ecma_Day(t) - ecma_DayFromYear(year));
+}
+
+static int* ecma_getfirstdays(int isleap) {
+  static int sdays[2][MonthsInYear+1] =
+    {{0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365},
+     {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366}};
+  
+  return sdays[isleap];
+}
+
+static int ecma_MonthFromTime(etime_t t, int year) {
+  int *days;
+  etimeint_t dwy = ecma_DayWithinYear(t, year);
+  
+  unsigned int i;
+  
+  days = ecma_getfirstdays(ecma_IsLeapYear(year));
+  
+  for(i = 0; i < MonthsInYear; i++) {
+    if(dwy >= days[i] && dwy < days[i+1]) {
+      return i;
+    }
+  }
+  
+  return -1;
+}
+
+static int ecma_DateFromTime(etime_t t, int year) {
+  int *days;
+  int mft = ecma_MonthFromTime(t, year);
+  int dwy = ecma_DayWithinYear(t, year);
+  
+  if(mft > 11) {
+    return -1;
+  }
+  
+  days = ecma_getfirstdays(ecma_IsLeapYear(year));
+  
+  return dwy - days[mft] + 1;
+}
+
+static int ecma_WeekDay(etime_t t) {
+  int ret = (ecma_Day(t)+4) % 7;
+  if(ret < 0) {
+    ret += 7;
+  }
+
+  return ret;
+}
+
+static int ecma_DaylightSavingTA(etime_t t) {
+  time_t time = t / 1000;
+  struct tm tm;
+  memset(&tm, 0, sizeof(t));
+  localtime_r(&time, &tm);
+  if(tm.tm_isdst > 0) {
+    return msPerHour;
+  } else {
+    return 0;
+  }
+}
+
+static int ecma_LocalTZA() {
+  return (int)-g_gmtoffms;
+}
+
+static etimeint_t ecma_LocalTime(etime_t t) {
+  return t + ecma_LocalTZA() + ecma_DaylightSavingTA(t);
+}
+
+static etimeint_t ecma_UTC(etime_t t) {
+  return t - ecma_LocalTZA() - ecma_DaylightSavingTA(t-ecma_LocalTZA());
+}
+
+#define DEF_EXCTRACT_TIMEPART(funcname, c1, c2) \
+  static int ecma_##funcname(etime_t t) { \
+    int ret = (etimeint_t)floor( t / c1) % c2; \
+    if(ret < 0) { \
+      ret += c2; \
+    } \
+    return ret; \
+  }
+
+DEF_EXCTRACT_TIMEPART(HourFromTime, msPerHour, HoursPerDay)
+DEF_EXCTRACT_TIMEPART(MinFromTime, msPerMinute, MinutesPerHour)
+DEF_EXCTRACT_TIMEPART(SecFromTime, msPerSecond, SecondsPerMinute)
+DEF_EXCTRACT_TIMEPART(msFromTime, 1, msPerSecond)
+
+
```

Comment:
remove unnecessary space.

Rationale: visual formatting in the code tends to be subjective and make it easy to introduce non-semantic diffs when people edit the code. (This is one of the reasons consistent style is important in general)